spring-cloud-deployer-resource-maven/src/main/java/org/springframework/cloud/deployer/resource/maven/MavenArtifactResolver.java

```diff
@@ -137,28 +137,39 @@ public MavenArtifactResolver(final MavenProperties properties) {
 								remoteRepository.getSnapshotPolicy().getUpdatePolicy(),
 								remoteRepository.getSnapshotPolicy().getChecksumPolicy()));
 			}
+			if (remoteRepositoryHasCredentials(remoteRepository)) {
+				final String username = remoteRepository.getAuth().getUsername();
+				final String password = remoteRepository.getAuth().getPassword();
+				remoteRepositoryBuilder.setAuthentication(newAuthentication(username, password));
+			}
+
+			RemoteRepository repo = remoteRepositoryBuilder.build();
+			Proxy proxy = null;
 			if (isProxyEnabled()) {
 				MavenProperties.Proxy proxyProperties = this.properties.getProxy();
-				if (this.authentication != null) {
-					remoteRepositoryBuilder.setProxy(new Proxy(
+				if (this.proxyAuthentication != null) {
+					proxy = new Proxy(
 							proxyProperties.getProtocol(),
 							proxyProperties.getHost(),
 							proxyProperties.getPort(),
-							this.authentication));
+							this.proxyAuthentication);
 				}
 				else {
 					// if proxy does not require authentication
-					remoteRepositoryBuilder.setProxy(new Proxy(
+					proxy = new Proxy(
 							proxyProperties.getProtocol(),
 							proxyProperties.getHost(),
-							proxyProperties.getPort()));
+							proxyProperties.getPort());
 				}
+
+				DefaultProxySelector proxySelector = new DefaultProxySelector();
+				proxySelector.add(proxy, this.properties.getProxy().getNonProxyHosts());
+
+				proxy = proxySelector.getProxy(repo);
 			}
-			if (remoteRepositoryHasCredentials(remoteRepository)) {
-				final String username = remoteRepository.getAuth().getUsername();
-				final String password = remoteRepository.getAuth().getPassword();
-				remoteRepositoryBuilder.setAuthentication(newAuthentication(username, password));
-			}
+
+			remoteRepositoryBuilder = new RemoteRepository.Builder(repo);
```

Comment:
Is there a reason why you create a builder again, then set a proxy and build again. Could this be done within `if` block where `proxy` is created?

Reply:
The RemoteRepository is readonly object. It can only be created from RemoteRepository.Builder. DefaultProxySelector only accept RemoteRepository as parameter. In order to utilize ProxySelector, I have to build RemoteRepository first, then call `proxySelector.add(proxy, this.properties.getProxy().getNonProxyHosts());`, then use the `public Builder( RemoteRepository prototype )` to recreate Builder, then build repository again.

The best way is to modify ProxySelector to accept, for example, repository host and protocol as parameter. Unfortunately, the eclipse aether project is already deprecated, it doesn't accept pull request anymore.

Reply:
Right that explains it then, thx. Btw eclipse project is defunct but whole project were moved into apache and afaik is still alive. https://github.com/apache/maven-resolver

Reply:
Hey @jvalkeal, thanks for the information. Please let me know if you want the change comes from eclipse aether. I'm not sure the change to eclipse aether would be accepted. It is not really a bug in aether. It would be feature proposal to aether.